lib/CodeGen/AsmPrinter/WinException.cpp

```diff
@@ -778,183 +778,298 @@ void WinException::emitExceptHandlerTable(const MachineFunction *MF) {
+    Left = FuncInfo.ClrEHUnwindMap[Left].Parent;
+    Right = FuncInfo.ClrEHUnwindMap[Right].Parent;
   }
 
   return Left;
 }
 
 const MCExpr *WinException::getOffset(MCSymbol *OffsetOf, MCSymbol *OffsetFrom) {
-  return  MCBinaryExpr::createSub(create32bitRef(OffsetOf), create32bitRef(OffsetFrom),
-                                  Asm->OutContext);
+  return MCBinaryExpr::createSub(create32bitRef(OffsetOf),
+                                 create32bitRef(OffsetFrom), Asm->OutContext);
 }
 
-void WinException::emitCLRExceptionTable(const MachineFunction *MF) {
-  const std::vector<LandingPadInfo> &PadInfos = MMI->getLandingPads();
-  const Function *Fn = MF->getFunction();
-  ClrEHFuncInfo &Info = MMI->getClrEHFuncInfo(Fn);
-
-  SmallVector<const LandingPadInfo *, 64> LandingPads;
-  LandingPads.reserve(PadInfos.size());
-  for (const auto &LP : PadInfos)
-    LandingPads.push_back(&LP);
-
-  // Compute label ranges for call sites as we would for the Itanium LSDA, but
-  // use an all zero action table because we aren't using these actions.
-  SmallVector<unsigned, 64> FirstActions;
-  FirstActions.resize(LandingPads.size());
-  SmallVector<CallSiteEntry, 64> CallSites;
-  computeCallSiteTable(CallSites, LandingPads, FirstActions);
-
-  MCSymbol *EHFuncBeginSym = Asm->getFunctionBegin();
-  MCSymbol *EHFuncEndSym = Asm->getFunctionEnd();
+const MCExpr *WinException::getOffset(const MCExpr *OffsetOf,
+                                      MCSymbol *OffsetFrom) {
+  return MCBinaryExpr::createSub(OffsetOf, create32bitRef(OffsetFrom),
+                                 Asm->OutContext);
+}
 
-  auto &StartEndLabelMap = Info.StartEndLabelMap;
-  StartEndLabelMap[Fn] = { EHFuncBeginSym, EHFuncEndSym };
-  bool IsLast = (StartEndLabelMap.size() == Info.Handlers.size() + 1);
+void WinException::emitCLRExceptionTable(const MachineFunction *MF) {
+  const int NullState = -1;
+  MCStreamer &OS = *Asm->OutStreamer;
+  const Function *F = MF->getFunction();
+  WinEHFuncInfo &FuncInfo = MMI->getWinEHFuncInfo(F);
+  MCSymbol *FuncBeginSym = Asm->getFunctionBegin();
+  MCSymbol *FuncEndSym = Asm->getFunctionEnd();
+
+  struct ClrClause {
+    MCSymbol *StartLabel;
+    MCSymbol *EndLabel;
+    int State;
+    int EnclosingState;
+  };
+  SmallVector<ClrClause, 8> Clauses;
+
+  // Build a map from handler MBBs to their corresponding states
+  int NumStates = FuncInfo.ClrEHUnwindMap.size();
+  DenseMap<const MachineBasicBlock *, int> HandlerStates;
+  for (int State = 0; State < NumStates; ++State) {
+    MachineBasicBlock *HandlerBlock =
+        FuncInfo.ClrEHUnwindMap[State].Handler.get<MachineBasicBlock *>();
+    HandlerStates[HandlerBlock] = State;
+  }
 
-  SmallVector<std::pair<MCSymbol *, HandlerTreeNode *>, 4> HandlerStack;
-  MCSymbol *LastLabel = nullptr;
+  // Write out a sentinel indicating the end of the xdata proper.
+  OS.EmitIntValue(0xffffffff, 4);
+  // Write out the number of funclets
+  OS.EmitIntValue(NumStates, 4);
```

Comment:
If there is no EH to report, is there some kind of early out we should take? Or do we never end up in this code to begin with?